### PR TITLE
Enable the client to configure from a connection URI - fixes #40

### LINKED ISSUE #40: Configure from URI
e.g : `postgresql://dbuser:secretpassword@database.server.com:3211/mydb`

### PR DESCRIPTION
**Pull request by BillyYccc**
Add feature for connecting with connection URI. Fixes https://github.com/reactiverse/reactive-pg-client/issues/40.

### CODE REVIEW

**Comment by BillyYccc** [comment]
Hi! I have reworked to implement based on regular expressions, I tidy the commit records and this PR is ready to be reviewed.

File: src/main/java/io/reactiverse/pgclient/impl/PgConnectionUriParser.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+  private static int NET_LOCATION_GROUP = 2;
+  private static int PORT_GROUP = 3;
+  private static int DATABASE_GROUP = 4;
+  private static int PARAMETER_GROUP = 5;
+
+  private final String connectionUri;
+
+  private JsonObject configuration = new JsonObject();
+
+  private PgConnectionUriParser(String connectionUri) {
```
Comment (vietj) [inline src/main/java/io/reactiverse/pgclient/impl/PgConnectionUriParser.java]: make everything static  and pass the object along the way in methods

**Comment by BillyYccc** [comment]
I doubt we should not provide configuration by 
`PgConnectOptions options = PgConnectionOptionsProvider.fromEnv();`
because the https://www.postgresql.org/docs/9.6/static/libpq-envars.html said that 
> The following environment variables can be used to select default connection parameter values, which will be used by PQconnectdb, PQsetdbLogin and PQsetdb **if no value is directly specified by the calling code.**


I think the priority is When options is specified, we use the specified one; if not specified, we use the env vars if the env vars exist, only when the option is not given and env vars do not exist, we use the default values finally.

**Comment by BillyYccc** [comment]
`PgConnectOptions options = PgConnectionOptionsProvider.fromUri(uri);`
This is a very appealing idea and I find this is also used in [libbq](https://www.postgresql.org/docs/9.6/static/libpq-connect.html).
> `PQconninfoOption *PQconninfoParse(const char *conninfo, char **errmsg);`

The connection string can either a connection uri like
 `postgresql://[user[:password]@][netloc][:port][/dbname][?param1=value1&...]` 
or a Keyword/Value Connection Strings e.g. 
` host=localhost port=5432 dbname=mydb connect_timeout=10`

So we can achieve one 
```
PgConnectOptions options = PgConnectionOptionsProvider.fromUri(uri);
PgConnectOptions options = PgConnectionOptionsProvider.fromKeyValues(kvString);
```

What do you think of this?

**Comment by vietj** [comment]
@BillyYccc I mean to set other settings like connection pool size, proxy, etc...

**Comment by vietj** [comment]
for now let's go static for the URI / keypairs

**Comment by vietj** [comment]
and also uses `PgConnectOptionsProvider` static methods

File: src/main/java/io/reactiverse/pgclient/impl/PgConnectionUriParser.java (hunk 1 above)
Comment (BillyYccc) [reply src/main/java/io/reactiverse/pgclient/impl/PgConnectionUriParser.java]: changes are in https://github.com/reactiverse/reactive-pg-client/pull/93/commits/92afd499d8a45971da47ebfb5ddf23fc505c176f.

**Comment by BillyYccc** [comment]
I also think we need a part `Configuration` on purpose in the documentation between the part `Getting started` and `Connecting to Postgres` to clarify the 3 ways of configuring:
1. setter 
2. uri 
3. env

**Comment by vietj** [comment]
yes but we can see that in another PR

**Comment by BillyYccc** [comment]
Do I need to add a same license like other files in the new file header ?

**Comment by vietj** [comment]
yes and you need to agree on the copyright terms

File: src/main/docs/index.md [hunk 2]
```diff
@@ -61,6 +61,12 @@ The following parameters are supported:
 * `PGUSER`
 * `PGPASSWORD`
 
+We also provide you an alternative way to connect when you want to configure with a connection URI:
```
Comment (vietj) [inline src/main/docs/index.md]: add a reference to the documentation that explains the connection string format

File: src/main/java/io/reactiverse/pgclient/PgClient.java [hunk 3]
```diff
@@ -55,6 +57,23 @@ static PgPool pool(Vertx vertx, PgPoolOptions options) {
     return new PgPoolImpl(vertx, false, options);
   }
 
+  /**
+   * Create a connection pool to the database configured with the given {@code connectionUri}.
+   *
+   * @param connectionUri the connection URI
+   * @return the connection pool
+   */
+  static PgPool pool(String connectionUri) {
```
Comment (vietj) [inline src/main/java/io/reactiverse/pgclient/PgClient.java]: we should not need both for the moment as user should do it itself using the provider

File: src/main/java/io/reactiverse/pgclient/PgClient.java [hunk 4]
```diff
@@ -86,6 +88,13 @@ static void connect(Vertx vertx, PgConnectOptions options, Handler<AsyncResult<P
     }
   }
 
+  /**
+   * Like {@link #connect(Vertx, PgConnectOptions, Handler)} but with a specific connection URI to configure the connection.
+   */
+  static void connect(Vertx vertx, String connectionUri, Handler<AsyncResult<PgConnection>> handler) {
```
Comment (vietj) [inline src/main/java/io/reactiverse/pgclient/PgClient.java]: we need to remove this for now

**Comment by vietj** [comment]
the invalid test should test catch exception, i.e invalid config should not return half configured options

**Comment by BillyYccc** [comment]
Now I feel let the parser returns a null when parsing invalid uris seems like a bad idea because the user doesn't know what happened(we need to at least give some warning), maybe throwing an IllegalArgumentException is a better choice? @vietj

**Comment by vietj** [comment]
looks nice, I'm going to merge it, thanks for the contribution!

**Comment by BillyYccc** [comment]
Feeling grateful for your guidance, Thank you!
I'm going to make another PR to improve the document later.

**Pull request merged**